pkg/cmd/server/origin/run_components.go

```diff
@@ -160,30 +163,52 @@ func (c *MasterConfig) RunServiceAccountTokensController(cm *cmapp.CMServer) {
+	controller := sacontroller.NewTokensController(
+		c.Informers.KubernetesInformers().Core().V1().ServiceAccounts().Informer(),
+		c.Informers.KubernetesInformers().Core().V1().Secrets().Informer(),
+		c.KubeClientsetExternal(),
+		options,
+	)
+	go controller.Run(int(cm.ConcurrentSATokenSyncs), utilwait.NeverStop)
 }
 
 // RunServiceAccountPullSecretsControllers starts the service account pull secret controllers
 func (c *MasterConfig) RunServiceAccountPullSecretsControllers() {
-	serviceaccountcontrollers.NewDockercfgDeletedController(c.KubeClientsetInternal(), serviceaccountcontrollers.DockercfgDeletedControllerOptions{}).Run()
-	serviceaccountcontrollers.NewDockercfgTokenDeletedController(c.KubeClientsetInternal(), serviceaccountcontrollers.DockercfgTokenDeletedControllerOptions{}).Run()
+	go serviceaccountcontrollers.NewDockercfgDeletedController(
+		c.Informers.InternalKubernetesInformers().Core().InternalVersion().Secrets().Informer(),
+		c.KubeClientsetInternal(),
+		serviceaccountcontrollers.DockercfgDeletedControllerOptions{},
+	).Run(utilwait.NeverStop)
+	go serviceaccountcontrollers.NewDockercfgTokenDeletedController(
+		c.Informers.InternalKubernetesInformers().Core().InternalVersion().Secrets().Informer(),
+		c.KubeClientsetInternal(),
+		serviceaccountcontrollers.DockercfgTokenDeletedControllerOptions{},
+	).Run(utilwait.NeverStop)
 
-	dockerURLsIntialized := make(chan struct{})
-	dockercfgController := serviceaccountcontrollers.NewDockercfgController(c.KubeClientsetInternal(), serviceaccountcontrollers.DockercfgControllerOptions{DockerURLsIntialized: dockerURLsIntialized})
+	dockerURLsInitialized := make(chan struct{})
+	dockercfgController := serviceaccountcontrollers.NewDockercfgController(
+		c.Informers.InternalKubernetesInformers().Core().InternalVersion().ServiceAccounts().Informer(),
+		c.Informers.InternalKubernetesInformers().Core().InternalVersion().Secrets().Informer(),
+		c.KubeClientsetInternal(),
+		serviceaccountcontrollers.DockercfgControllerOptions{DockerURLsInitialized: dockerURLsInitialized},
+	)
 	go dockercfgController.Run(5, utilwait.NeverStop)
 
 	dockerRegistryControllerOptions := serviceaccountcontrollers.DockerRegistryServiceControllerOptions{
-		RegistryNamespace:    "default",
-		RegistryServiceName:  "docker-registry",
-		DockercfgController:  dockercfgController,
-		DockerURLsIntialized: dockerURLsIntialized,
+		RegistryNamespace:     "default",
+		RegistryServiceName:   "docker-registry",
+		DockercfgController:   dockercfgController,
+		DockerURLsInitialized: dockerURLsInitialized,
 	}
-	go serviceaccountcontrollers.NewDockerRegistryServiceController(c.KubeClientsetInternal(), dockerRegistryControllerOptions).Run(10, make(chan struct{}))
+	go serviceaccountcontrollers.NewDockerRegistryServiceController(
+		c.Informers.InternalKubernetesInformers().Core().InternalVersion().Secrets().Informer(),
```

Comment:
It's not needed, we only have a single item we are watching, and it's less efficient to filter the entire informer set for one items